examples/threaded_example/threaded.rs

```diff
@@ -0,0 +1,529 @@
+
+    Ok(())
+}
+
+/// Runs the specified function `f`, returning either an error (if `f` failed)
+/// or the execution time of `f`.
+fn measure<T>(mut f: impl FnMut() -> anyhow::Result<T>) -> anyhow::Result<Duration> {
+    let now = Instant::now();
+    f()?;
+    Ok(now.elapsed())
+}
+
+/// Helper type for running sub-protocols for the coordinator.
+struct SubProtocolRunner {
+    /// Channels to send messages to our workers.
+    send_to_workers: WorkerChannels,
+    /// Receive messages from worker threads to route to other workers.
+    from_workers: Receiver<MessageFromWorker>,
+    /// Total number of Participants/worker threads participating in our
+    /// sub-protocol.
+    total_workers: usize,
+    /// Execution time data gathered during execution.
+    statistics: Statistics,
+}
+
+impl SubProtocolRunner {
+    /// Initialize `SubProtocolRunner` with the given channels for sending
+    /// messages to worker threads.
+    pub fn new(
+        send_to_workers: WorkerChannels,
+        from_workers: Receiver<MessageFromWorker>,
+        total_workers: usize,
+    ) -> Self {
+        Self {
+            send_to_workers,
+            from_workers,
+            total_workers,
+            statistics: Statistics::new(),
+        }
+    }
+
+    /// Runs an entire sub-protocol from start to end. Records statistics about
+    /// sub-protocol execution time from main thread's point of view.
+    fn do_sub_protocol(&mut self, sub_protocol: SubProtocol, key_id: KeyId) -> anyhow::Result<()> {
+        info!(
+            "Starting sub-protocol: {:?} for {:?}.",
+            sub_protocol, key_id
+        );
+        let duration = measure(|| {
+            self.start_new_subprotocol(sub_protocol, key_id)?;
+            self.route_worker_messages()?;
+            Ok(())
+        })?;
+        self.record(sub_protocol, duration);
+
+        info!("Finished sub-protocol: {:?}.", sub_protocol);
+        Ok(())
+    }
+
+    /// (Helper Method) Start the specified `sub_protocol` by sending a message
+    /// to all worker threads.
+    fn start_new_subprotocol(
+        &self,
+        sub_protocol: SubProtocol,
+        key_id: KeyId,
+    ) -> anyhow::Result<()> {
+        debug!("Starting sub-protocol: {:?}", sub_protocol);
+        let rng = &mut thread_rng();
+        let sid = SessionId(Identifier::random(rng));
+
+        for worker in self.send_to_workers.values() {
+            worker.send(MessageFromCoordinator::NewSubProtocol(
+                sub_protocol,
+                key_id,
+                sid,
+            ))?;
+        }
+
+        Ok(())
+    }
+
+    /// (Helper Method) Receives messages from workers via our
+    /// `self.from_workers` field and routes messages to the correct worker
+    /// thread via our `self.send_to_workers` map.
+    fn route_worker_messages(&self) -> anyhow::Result<()> {
+        // # of workers who finished the current sub_protocol.
+        let mut sub_protocol_ended = 0;
```

Comment:
todo: We discussed on a call that this singular output counter means that the coordinator itself doesn't support concurrent protocol executions, even though the workers do. Please make a new issue about adding coordinator support for concurrency.